### PR TITLE
[DEV-645] - Storage refactor

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
I closed the other refactor PR because it had tons of conflicts after the recent merge. In this PR I separated commits into smaller pieces. The first commit is moving the storage variables from `Badges.sol` into `BadgesStorage.sol`. The second moving utility-like functions out of `Badges.sol` into `Utils.sol`.

Along with this PR I made a new PR (https://github.com/otterspace-xyz/otterspace-contracts/pull/104) for improving the way that we test the contracts. That PR should be merged first, then we can run that code on this branch (since it's a necessary dependency)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/BadgeStorage.sol [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import { SignatureCheckerUpgradeable } from "@openzeppelin-upgradeable/utils/cryptography/SignatureCheckerUpgradeable.sol";
+import { BitMaps } from "@openzeppelin/contracts/utils/structs/BitMaps.sol";
+import { OwnableUpgradeable } from "@openzeppelin-upgradeable/access/OwnableUpgradeable.sol";
+import { EIP712Upgradeable } from "@openzeppelin-upgradeable/utils/cryptography/draft-EIP712Upgradeable.sol";
+import { UUPSUpgradeable } from "@openzeppelin-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { ERC165Upgradeable } from "@openzeppelin-upgradeable/utils/introspection/ERC165Upgradeable.sol";
+import { IERC721Metadata } from "./interfaces/IERC721Metadata.sol";
+import { ISpecDataHolder } from "./interfaces/ISpecDataHolder.sol";
+
+contract BadgeStorage is ERC165Upgradeable, UUPSUpgradeable, OwnableUpgradeable, EIP712Upgradeable {
```
Comment (rahulrumalla) [inline src/BadgeStorage.sol:14]: First impression here is that we're used suffix `DataHolder` naming for Spec and `Storage` here. Would it make sense to be more consistent or do u see them as separate things?

File: src/Utils.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  /**
+   * @notice Allows the Badges contract to communicate with the SpecDataHolder contract
+   * @param _dataHolder address of the SpecDataHolder contract
+   */
+  function setDataHolder(address _dataHolder) external virtual onlyOwner {
+    specDataHolder = ISpecDataHolder(_dataHolder);
+  }
+
+  function name() external view virtual override returns (string memory) {
```
Comment (rahulrumalla) [inline src/Utils.sol:61]: name, symbol and tokenURI are actually for `EIP721Metadata` compatibility right? Should we perhaps modularize that into it's own things? It's not really a general purpose utility to belong in "Utils.sol" I feel

File: src/Utils.sol [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+pragma solidity 0.8.16;
+import { BadgeStorage } from "./BadgeStorage.sol";
+import { SignatureCheckerUpgradeable } from "@openzeppelin-upgradeable/utils/cryptography/SignatureCheckerUpgradeable.sol";
+import { ISpecDataHolder } from "./interfaces/ISpecDataHolder.sol";
+import { IERC721Metadata } from "./interfaces/IERC721Metadata.sol";
+
+bytes32 constant AGREEMENT_HASH = keccak256("Agreement(address active,address passive,string tokenURI)");
+
+contract Utils is IERC721Metadata, BadgeStorage {
+  function getBadgeIdHash(address _to, string memory _uri) public view virtual returns (bytes32) {
```
Comment (rahulrumalla) [inline src/Utils.sol:11]: Why does this belong here in Utils?

File: src/Badges.sol [hunk 4]
```diff
@@ -19,27 +19,9 @@ contract Badges is
   ERC165Upgradeable,
```
Comment (rahulrumalla) [inline src/Badges.sol:19]: Are these deps redundant as BadgesStorage already inherit it?

**Review by rahulrumalla — CHANGES_REQUESTED** [review]

File: src/Utils.sol (hunk 2 above)
Comment (grp06) [reply src/Utils.sol:61]: Yeah true. For now I put it back in Badges.sol

File: src/Utils.sol (hunk 3 above)
Comment (grp06) [reply src/Utils.sol:11]: My thought was that it felt like some kind of helper function but I guess it's pretty specific to the minting function and not generalizable

File: src/Badges.sol (hunk 4 above)
Comment (grp06) [reply src/Badges.sol]: yup. fixed

**Pull request closed without merging**